File: Task/Kosaraju/Visual-Basic-.NET/kosaraju.vb
Module Module1

    Function Kosaraju(g As List(Of List(Of Integer))) As List(Of Integer)
        Dim size = g.Count
        Dim vis(size - 1) As Boolean
        Dim l(size - 1) As Integer
        Dim x = size

        Dim t As New List(Of List(Of Integer))
        For i = 1 To size
            t.Add(New List(Of Integer))
        Next

        Dim visit As Action(Of Integer) = Sub(u As Integer)
                                              If Not vis(u) Then
                                                  vis(u) = True
                                                  For Each v In g(u)
                                                      visit(v)
                                                      t(v).Add(u)
                                                  Next
                                                  x -= 1
                                                  l(x) = u
                                              End If
                                          End Sub

        For i = 1 To size
            visit(i - 1)
        Next
        Dim c(size - 1) As Integer

        Dim assign As Action(Of Integer, Integer) = Sub(u As Integer, root As Integer)
                                                        If vis(u) Then
                                                            vis(u) = False
                                                            c(u) = root
                                                            For Each v In t(u)
                                                                assign(v, root)
                                                            Next
                                                        End If
                                                    End Sub

        For Each u In l
            assign(u, u)
        Next

        Return c.ToList
    End Function

    Sub Main()
        Dim g = New List(Of List(Of Integer)) From {
            New List(Of Integer) From {1},
            New List(Of Integer) From {2},
            New List(Of Integer) From {0},
            New List(Of Integer) From {1, 2, 4},
            New List(Of Integer) From {3, 5},
            New List(Of Integer) From {2, 6},
            New List(Of Integer) From {5},
            New List(Of Integer) From {4, 6, 7}
        }

        Dim output = Kosaraju(g)
        Console.WriteLine("[{0}]", String.Join(", ", output))
    End Sub

End Module

File: Task/Kosaraju/PascalABC.NET/kosaraju.pas
function kosaraju(g: List<Array of integer>): array of integer;
var
  size := g.Count;
  vis: array of boolean := new boolean[size];
  l: array of integer := new integer[size];
  c: array of integer := new integer[size];
  x := size;
  t: array of List<integer> := new List<integer>[size];

  procedure visit(u: integer);
  begin
    if not vis[u] then
    begin
      vis[u] := true;
      foreach var v in g[u] do
      begin
        visit(v);
        t[v].Add(u);
      end;
      dec(x);
      l[x] := u;
    end;
  end;

  procedure assign(u, root: integer);
  begin
    if vis[u] then
    begin
      vis[u] := false;
      c[u] := root;
      foreach var v in t[u] do assign(v, root);
    end;
  end;

begin
  for var i := 0 to t.Count - 1 do t[i] := new List<integer>;

  foreach var u in [0..size - 1] do visit(u);
  foreach var u in l do assign(u, u);
  result := c;
end;

begin
  var g := Lst(|1|, |2|, |0|, |1, 2, 4|, |3, 5|, |2, 6|, |5|, |4, 6, 7|);
  println(kosaraju(g));
end.
